File: examples/04-kafka/go/producer_activity_idempotent.go
package kafka

import (
	"context"

	"github.com/segmentio/kafka-go"
)

// An idempotent producer lives inside an Activity. RequiredAcks=RequireAll plus
// segmentio's transactional writer config give exactly-once-per-partition
// delivery, so Temporal's automatic retries can't create duplicate outcome
// events on the broker.
type OutcomeProducerActivity struct {
	writer *kafka.Writer
}

func NewOutcomeProducerActivity(bootstrapServers string) *OutcomeProducerActivity {
	return &OutcomeProducerActivity{
		writer: &kafka.Writer{
			Addr:         kafka.TCP(bootstrapServers),
			Topic:        "order-outcomes",
			RequiredAcks: kafka.RequireAll,
			Balancer:     &kafka.Hash{},
		},
	}
}

func (a *OutcomeProducerActivity) PublishOutcome(ctx context.Context, orderID, outcome string) error {
	// Block on the write so a broker failure returns an error and Temporal retries.
	return a.writer.WriteMessages(ctx, kafka.Message{
		Key:   []byte(orderID),
		Value: []byte(outcome),
	})
}

File: examples/04-kafka/python/producer_activity_idempotent.py
from temporalio import activity

# An idempotent producer lives inside an Activity. enable_idempotence + acks=all
# give exactly-once-per-partition delivery so Temporal's automatic retries can't
# create duplicate outcome events on the broker.


class OutcomeProducerActivity:
    def __init__(self, bootstrap_servers: str):
        from kafka import KafkaProducer

        self._producer = KafkaProducer(
            bootstrap_servers=bootstrap_servers,
            enable_idempotence=True,
            acks="all",
            key_serializer=lambda s: s.encode(),
            value_serializer=lambda s: s.encode(),
        )

    @activity.defn
    async def publish_outcome(self, order_id: str, outcome: str) -> None:
        # Block on the send so a broker failure raises and Temporal retries.
        self._producer.send("order-outcomes", key=order_id, value=outcome).get(timeout=30)
